### PR TITLE
Add useSettings hook for reading multiple settings at once

### PR DESCRIPTION
**Pull request by jsnajdr**
This is an attempt to improve the performance situation discussed in #54819, to make typing faster. I expected better results, but here it is anyway.

The problem is that 1000 blocks create 80k subscriptions to the `core/block-editor` store, and every action dispatch leads to calling 80k selectors and checking if their return value has changed. Of course, in a vast majority of cases the return value hasn't changed.

`useSelect` has async mode that's active on almost all of the 1000 blocks, but even adding an item to the priority queue 80k times takes some effort.

This PR addresses one hook that is to blame for many (20k out of 80k) of the store subscriptions: `useSetting`.

The first few commits perform optimizations on the existing `useSetting` hook, making it in total 40% faster. Don't create new strings, memoize splitting `a.b.c` paths into `['a', 'b', 'c']` arrays (`lodash.get` always did this, too), etc.

The second part of the PR is creating a new hook, `useSettings`. Instead of reading settings individually:
```js
const allowFixed = useSetting( 'position.fixed' );
const allowSticky = useSetting( 'position.sticky' );
```
we merge them into one hook call:
```js
const [ allowFixed, allowSticky ] = useSettings( [ 'position.fixed', 'position.sticky' ] );
```
The point is that each `useSetting` hook has one `useSelect` inside, leading to one store subscription per hook. My merging the setting reads together, we can save a lot of subscriptions.

Migrating most usages of `useSetting` to `useSettings` reduced the number of `block-editor` subscription, on a large post with 1000 blocks, from 80k to 60k.

The improvements are nice but still very incremental, they don't really solve the problem. We'll need to continue to search for other solutions.

The last commit contains some performance-measuring debug code I've been using. There are:
- `console.log` statements that report the number of `block-editor` subscriptions
- `console.log` statements that report on how the async mode priority queue is being processed in a series of `requestIdleCallback`s. Gives a good picture on how many store listeners we can process per second.

I'll remove this debugging code before merging (hopefully 🙂)

This PR is best tested with a 4x CPU throttling enabled in the Performance tab in Chrome devtools.

### CODE REVIEW

File: packages/block-editor/src/hooks/utils.js [hunk 1]
```diff
@@ -126,48 +126,93 @@ export function shouldSkipSerialization( blockType, featureSet, feature ) {
  * @return {Object} Settings object.
  */
 export function useBlockSettings( name, parentLayout ) {
```
Comment (youknowriad) [inline packages/block-editor/src/hooks/utils.js:128]: For this particular function, my goal was to actually move its call to `BlockEdit` or something and add it to the context of a block (probably as a private thing), that way it's only called once per block (as opposed to one per hook like today).

**Comment by jsnajdr** [comment]
I removed the commit with debugging code to make the CI check pass. You can manually apply it with `git cherry-pick  c029303`.

File: packages/block-editor/src/utils/object.js [hunk 2]
```diff
@@ -125,7 +128,7 @@ export function setImmutably( object, path, value ) {
  * @return {*} Value of the object property at the specified path.
  */
 export const getValueFromObjectPath = ( object, path, defaultValue ) => {
-	const normalizedPath = Array.isArray( path ) ? path : path.split( '.' );
+	const normalizedPath = Array.isArray( path ) ? path : stringToPath( path );
```
Comment (tyxla) [inline packages/block-editor/src/utils/object.js:131]: This can totally be shipped as a separate PR regardless of how we decide to move forward.

File: packages/block-editor/src/components/use-setting/index.js [hunk 3]
```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
...
  * It looks up the settings first in the block instance hierarchy.
  * If none is found, it'll look it up in the block editor store.
  *
- * @param {string} path The path to the setting.
- * @return {any} Returns the value defined for the setting.
- * @example
- * ```js
- * const isEnabled = useSetting( 'typography.dropCap' );
- * ```
+ * @param {string[]} paths The path to the setting.
```
Comment (tyxla) [inline packages/block-editor/src/components/use-setting/index.js]:
```suggestion
 * @param {string[]} paths The paths to the settings.
```

File: packages/block-editor/src/components/use-setting/index.js [hunk 4]
```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
...
+		result = [ 'default', 'theme', 'custom' ].flatMap(
+			( key ) => value[ key ] ?? []
+		);
+		mergeCache.set( value, result );
+	}
+	return result;
+}
+
 /**
  * Hook that retrieves the given setting for the block instance in use.
```
Comment (tyxla) [inline packages/block-editor/src/components/use-setting/index.js]:
```suggestion
 * Hook that retrieves the given settings for the block instance in use.
```

File: packages/block-editor/src/components/use-setting/index.js [hunk 5]
```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
...
  *
  * It looks up the settings first in the block instance hierarchy.
  * If none is found, it'll look it up in the block editor store.
```
Comment (tyxla) [inline packages/block-editor/src/components/use-setting/index.js]:
```suggestion
 * If none are found, it'll look them up in the block editor store.
```

File: packages/block-editor/src/components/use-setting/index.js [hunk 6]
```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
...
- * @example
- * ```js
- * const isEnabled = useSetting( 'typography.dropCap' );
- * ```
+ * @param {string[]} paths The path to the setting.
+ * @return {any[]} Returns the values defined for the settings.
  */
-export default function useSetting( path ) {
-	const { name: blockName, clientId } = useBlockEditContext();
+export function useSettings( paths ) {
```
Comment (tyxla) [inline packages/block-editor/src/components/use-setting/index.js]: Does it make sense to move `useSettings()` to a separate file? Perhaps under `../use-settings/index.js`? Right now it's a bit unfortunate that `use-setting/index.js` exports `useSettings()`.

File: packages/block-editor/README.md [hunk 7]
```diff
@@ -962,6 +962,20 @@ _Returns_
...
+
+It looks up the settings first in the block instance hierarchy. If none is found, it'll look it up in the block editor store.
+
+_Parameters_
+
+-   _paths_ `string[]`: The path to the setting.
+
+_Returns_
+
+-   `any[]`: Returns the values defined for the settings.
```
Comment (tyxla) [inline packages/block-editor/README.md]:
```suggestion
Hook that retrieves the given settings for the block instance in use.

It looks up the settings first in the block instance hierarchy. If none are found, it'll look them up in the block editor store.

_Parameters_

-   _paths_ `string[]`: The paths to the settings.

_Returns_

-   `any[]`: Returns the values defined for the settings.
```

File: packages/block-editor/src/components/font-sizes/font-size-picker.js [hunk 8]
```diff
@@ -6,17 +6,19 @@ import { FontSizePicker as BaseFontSizePicker } from '@wordpress/components';
 /**
  * Internal dependencies
  */
-import useSetting from '../use-setting';
+import { useSettings } from '../use-setting';
 
 function FontSizePicker( props ) {
-	const fontSizes = useSetting( 'typography.fontSizes' );
-	const disableCustomFontSizes = ! useSetting( 'typography.customFontSize' );
+	const [ fontSizes, customFontSizes ] = useSettings( [
```
Comment (tyxla) [inline packages/block-editor/src/components/font-sizes/font-size-picker.js]: Maybe we should keep the same name as the subsetting name:

```suggestion
	const [ fontSizes, customFontSize ] = useSettings( [
```

File: packages/block-editor/src/components/font-sizes/font-size-picker.js [hunk 9]
```diff
@@ -6,17 +6,19 @@ import { FontSizePicker as BaseFontSizePicker } from '@wordpress/components';
...
+		'typography.fontSizes',
+		'typography.customFontSize',
+	] );
 
 	return (
 		<BaseFontSizePicker
 			{ ...props }
 			fontSizes={ fontSizes }
-			disableCustomFontSizes={ disableCustomFontSizes }
+			disableCustomFontSizes={ ! customFontSizes }
```
Comment (tyxla) [inline packages/block-editor/src/components/font-sizes/font-size-picker.js]:
```suggestion
			disableCustomFontSizes={ ! customFontSize }
```

File: packages/block-editor/src/hooks/utils.js [hunk 10]
```diff
@@ -126,48 +126,93 @@ export function shouldSkipSerialization( blockType, featureSet, feature ) {
...
+		themeGradientPalette,
+		defaultGradientPalette,
+		defaultGradients,
+		areCustomGradientsEnabled,
+		isBackgroundEnabled,
+		isLinkEnabled,
+		isTextEnabled,
+		isHeadingEnabled,
+		isButtonEnabled,
+	] = useSettings( [
```
Comment (tyxla) [inline packages/block-editor/src/hooks/utils.js]: I think it could be nice to consider building an ESLint rule that prevents folks from having multiple `useSetting` calls at the same level, recommending that we use `useSettings()` instead.

Not for this PR, obviously, but something that we discussed earlier and I wanted to make sure it doesn't fall through the cracks.

**Review by tyxla — COMMENTED** [review]
Nice work 🥇 

Thanks for diving deeper into the large-post typing performance issues. 👍 

If `useSettings` is preferred for multiple settings, I think we should ensure to educate folks about it (adding some additional explanation in docs) and maybe introduce an ESLint rule that prevents sibling `useSetting()` calls?

File: packages/block-editor/src/components/font-sizes/font-size-picker.js (hunk 8 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/font-sizes/font-size-picker.js]: Here I chose the variable name based on the name of the setting: `customFontSize`.

File: packages/block-editor/src/components/use-setting/index.js (hunk 6 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-setting/index.js]: The hooks are very closely related, it's not justified to have them in separate files.

Most consumers should import them from the parent `components` module, not using the `/use-setting` subpath.

File: packages/block-editor/src/components/use-setting/index.js (hunk 5 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-setting/index.js]: Pushed a commit that fixes all the typos in the docblocks.

File: packages/block-editor/src/components/font-sizes/font-size-picker.js (hunk 8 above)
Comment (tyxla) [reply packages/block-editor/src/components/font-sizes/font-size-picker.js]: I may be missing something, but that's exactly why I added the suggestion - in the PR we're using `customFontSizes` and not `customFontSize`.

File: packages/block-editor/src/components/use-setting/index.js (hunk 6 above)
Comment (tyxla) [reply packages/block-editor/src/components/use-setting/index.js]: I disagree because I still find it confusing and unexpected, but I'll agree to disagree and commit.

File: packages/block-editor/src/components/use-setting/index.js [hunk 11]
```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
...
- * ```
+ * @param {string[]} paths The paths to the settings.
+ * @return {any[]} Returns the values defined for the settings.
  */
-export default function useSetting( path ) {
-	const { name: blockName, clientId } = useBlockEditContext();
+export function useSettings( paths ) {
+	const { name: blockName, clientId = null } = useBlockEditContext();
+
+	paths = useMemo( () => paths, paths );
```
Comment (dmsnell) [inline packages/block-editor/src/components/use-setting/index.js]: what's the purpose of this `useMemo`? am I mistaken in that it's doing nothing since all the internal function does is return its stored value?

File: packages/block-editor/src/components/use-setting/index.js [hunk 12]
```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
...
+function mergeOrigins( value ) {
+	let result = mergeCache.get( value );
+	if ( ! result ) {
+		result = [ 'default', 'theme', 'custom' ].flatMap(
+			( key ) => value[ key ] ?? []
+		);
+		mergeCache.set( value, result );
+	}
+	return result;
+}
```
Comment (dmsnell) [inline packages/block-editor/src/components/use-setting/index.js]: it would be nice to have a docblock-style comment explaining the purpose of this function, and I wonder if the name could be clearer since `mergeOrigins` doesn't outwardly communicate that there's a cache involved.

I'm also confused on the value we're storing here, or why we need `flatMap()`. In the old code it's not clear to me either, so maybe this is why 🤷‍♂️ but it looks like it might be using `.concat()` with a default of `[]` to avoid needing a conditional for the case where `result` is missing the necessary key. I guess it's convenient for when keys are undefined and then it automatically returns an array with only those values.

but the old code was directly returning these things too while here we're storing them. since these are retrieved from `value` is there a need to split them out at this point? did you find that extracting these three properties was showing up in the hot path?

I'm so confused on all of this, really. is it worth adding a separate function and cache vs. directly building the array every time? maybe it was all confused by the nature of the original use of `reduce()`? I'm not trying to argue that the cache isn't worth it; I'm only asking if we have reason to believe it is.

```js
const values = [];
[ 'default', 'theme', 'custom' ].forEach(
	( key ) => {
		if ( value[ key ] ) {
			values.push( value[ key ] );
		}
	}
)

return values;
```

File: packages/block-editor/src/components/use-setting/index.js [hunk 13]
```diff
@@ -93,126 +94,158 @@ const removeCustomPrefixes = ( path ) => {
...
- * @example
- * ```js
- * const isEnabled = useSetting( 'typography.dropCap' );
- * ```
+ * @param {string[]} paths The paths to the settings.
+ * @return {any[]} Returns the values defined for the settings.
  */
-export default function useSetting( path ) {
-	const { name: blockName, clientId } = useBlockEditContext();
+export function useSettings( paths ) {
```
Comment (dmsnell) [inline packages/block-editor/src/components/use-setting/index.js]: did you consider using variable argument counts here? it seems like most cases we'd be calling this we'll have literal values vs. dynamically-created arrays and I wonder if that's a convenience that would be nice.

```js
const [ … ] = useSettings( 'setting-a', 'setting-b', 'color.setting.thing' );
```

**Review by dmsnell — COMMENTED** [review]
very nice update here, and one that is probably overdue. thanks for addressing this - the reduction in hook count is great.

File: packages/block-editor/src/components/use-setting/index.js (hunk 11 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-setting/index.js]: On each `useSettings` call, the `paths` array will be different, although it contains the same values. This `useMemo` is to ensure that `paths` is a constant in that case, and to ensure that `useSelect` can memoize its return value. `useSelect` itself does something similar internally.

File: packages/block-editor/src/components/use-setting/index.js (hunk 13 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-setting/index.js]: Yes, that's a good idea. Variable arguments are better than an array argument. This is quite related to @tyxla's feedback about the module name:

> Does it make sense to move `useSettings()` to a separate file? Perhaps under `../use-settings/index.js`? Right now it's a bit unfortunate that `use-setting/index.js` exports `useSettings()`.

To solve both problems, we could do this:
1. Rewrite `useSettings` to use variable arguments.
2. Migrate all remaining usages of `useSetting` to `useSettings`. The entire difference is the return value as array: `const [ a ] = useSettings( 'a' )` vs `const a = useSetting( 'a' )`.
3. This makes `useSetting` practically deprecated: there's no good reason any more to use it instead of `useSettings`.
4. The actual source file can then be renamed to `use-settings/index.js`. The main thing it exports is `useSettings`, and then there's also the little `useSetting` back-compat wrapper. I think that solves the naming concerns very well.

Thoughts?

Comment (tyxla) [reply packages/block-editor/src/components/use-setting/index.js]: I really like that idea 👍 one fewer new API while still supporting all original use cases. Any good reason not to go that way?

File: packages/block-editor/src/components/use-setting/index.js (hunk 11 above)
Comment (dmsnell) [reply packages/block-editor/src/components/use-setting/index.js]: aha! the second value is the dependencies array, which in this case is `paths` itself. in my head I was reading `[paths]`

File: packages/block-editor/src/components/use-setting/index.js (hunk 13 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-setting/index.js]: Implemented! All usages of `useSetting` are refactored to `useSettings`, and the source folder name is now `use-settings`.

File: packages/block-editor/src/components/font-sizes/font-size-picker.js (hunk 8 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/font-sizes/font-size-picker.js]: Ah, I see 🙂 Changed as suggested.

File: packages/block-editor/src/components/use-setting/index.js (hunk 12 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-setting/index.js]: Some settings, like `color.palette`, have a "raw" value where settings from different "origins" are in separate arrays:
```js
{
  default: [ ... ],
  theme: [ ... ],
  custom: [ ... ],
}
```
The `mergeOrigins` function merges the three arrays into one, concatenating them together. I added memoization so that two calls to `useSelect` with the same state and arguments return the same value. This was a little pre-existing bug.

I added a docblock to the function that contains this explanation.

**Comment by jsnajdr** [comment]
@tyxla @dmsnell I believe I addressed all feedback, now ready for final review.

File: packages/block-library/src/group/edit.js [hunk 14]
```diff
@@ -98,11 +97,7 @@ function GroupEdit( {
 	} = attributes;
 
 	// Layout settings.
-	const defaultLayout = useSetting( 'layout' ) || {};
-	const usedLayout = ! layout?.type
-		? { ...defaultLayout, ...layout, type: 'default' }
-		: { ...defaultLayout, ...layout };
-	const { type = 'default' } = usedLayout;
```
Comment (jsnajdr) [inline packages/block-library/src/group/edit.js:105]: @tellthemachines This is a change that could be a separate PR. I found that the `useSetting( 'layout' )` call can be completely removed, because the `usedLayout` value is no longer used. Only its `.type` field is used. And the `type` can be calculated without knowing `defaultLayout`. That value always comes from the `layout` object.

`usedLayout` was used until #47477 before usages of `__experimentalLayout`, but today it's redundant.

**Comment by dmsnell** [comment]
@jsnajdr a couple questions on deprecation, you may have addressed and I may have missed.

 - should we add a deprecation notice to `useSetting`?
 - should we leave in a wrapping stub in `use-setting/index.js`?

I wasn't sure if any imports of `useSetting` change with the rename of the folder, but I think existing imports should continue to work. only wanted to double check that we don't suddenly break third-party code on this.

that being said, if we want to move away from `useSetting` maybe it would be good to call `deprecated()` in there and start pushing people towards `useSettings`?

File: packages/block-library/src/group/edit.js (hunk 14 above)
Comment (tellthemachines) [reply packages/block-library/src/group/edit.js:105]: Ah well spotted! Yes, this can be removed. Not too fussed about whether it's done here or in a separate PR, your choice 😄

**Comment by jsnajdr** [comment]
> should we add a deprecation notice to `useSetting`?

I wasn't sure if I want to really formally deprecate `useSetting`, with warnings at all, but why not? I added a commit that does that.

> should we leave in a wrapping stub in `use-setting/index.js`?

I think by now the names of the folders and files are sufficiently fine. The main thing that `use-settings/index.js` contains is the `useSettings` function, and the names correspond. Additionally it exports the legacy `useSetting`. It doesn't need to be a separate module.

> I wasn't sure if any imports of useSetting change with the rename of the folder

This is just the internal layout of the `block-editor`package, and internal imports. The public API is exported by the root `@wordpress/block-editor`. And also by the `/components` sub-module, but that's also only package-internal.

File: packages/block-editor/CHANGELOG.md [hunk 15]
```diff
@@ -2,21 +2,23 @@
 
 ## Unreleased
 
+-   Deprecated the `useSetting` function in favor of new `useSettings` one that can retrieve multiple settings at once.
```
Comment (tyxla) [inline packages/block-editor/CHANGELOG.md]:
```suggestion
-   Deprecated the `useSetting` function in favor of new `useSettings` one that can retrieve multiple settings at once ([#55337](https://github.com/WordPress/gutenberg/pull/55337)).
```

File: packages/block-editor/README.md [hunk 16]
```diff
@@ -944,9 +944,11 @@ _Parameters_
 
 ### useSetting
 
+> **Deprecated**
+
```
Comment (tyxla) [inline packages/block-editor/README.md:948]: Should we add a note right here that it's deprecated in favor of `useSettings()`?

File: packages/block-editor/src/components/inner-blocks/index.js [hunk 17]
```diff
@@ -98,7 +98,7 @@ function UncontrolledInnerBlocks( props ) {
 
 	const { allowSizingOnChildren = false } = defaultLayoutBlockSupport;
 
-	const defaultLayout = useSetting( 'layout' ) || EMPTY_OBJECT;
+	const [ defaultLayout ] = useSettings( 'layout' );
```
Comment (tyxla) [inline packages/block-editor/src/components/inner-blocks/index.js:101]: Should we keep the `EMPTY_OBJECT` fallback here?

File: packages/block-editor/src/components/spacing-sizes-control/hooks/use-spacing-sizes.js [hunk 18]
```diff
@@ -6,13 +6,15 @@ import { __ } from '@wordpress/i18n';
...
 export default function useSpacingSizes() {
-	const spacingSizes = [
-		{ name: 0, slug: '0', size: 0 },
-		...( useSetting( 'spacing.spacingSizes' ) || [] ),
-	];
+	const spacingSizes = [ { name: 0, slug: '0', size: 0 } ];
+
+	const [ settingsSizes ] = useSettings( 'spacing.spacingSizes' );
+	if ( settingsSizes ) {
+		spacingSizes.push( ...settingsSizes );
```
Comment (tyxla) [inline packages/block-editor/src/components/spacing-sizes-control/hooks/use-spacing-sizes.js:16]: Would be nice to avoid the mutation here.

File: packages/block-editor/src/components/use-settings/index.js [hunk 19]
```diff
@@ -0,0 +1,272 @@
...
+					// eslint-disable-next-line no-console
+					console.warn(
+						'Top level useSetting paths are disabled. Please use a subpath to query the information needed.'
+					);
+					return undefined;
+				}
+
+				// 0. Allow third parties to filter the block's settings at runtime.
+				let result = applyFilters(
+					'blockEditor.useSetting.before',
```
Comment (tyxla) [inline packages/block-editor/src/components/use-settings/index.js:170]: There's no need to introduce a bulk settings filter right?

File: packages/block-editor/src/components/use-settings/README.md [hunk 20]
```diff
@@ -1,8 +1,8 @@
-## Use Setting
+## Use Settings
 
-`useSetting` is a hook that will retrive the setting for the block instance that's in use. 
+`useSettings` is a hook that will retrieve the settings for the block instance that's in use.
```
Comment (tyxla) [inline packages/block-editor/src/components/use-settings/README.md:3]: Should we keep a reference to the deprecated `useSetting()` in the README?

File: packages/block-editor/src/components/use-settings/index.js [hunk 21]
```diff
@@ -0,0 +1,272 @@
...
+
+/**
+ * Hook that retrieves the given setting for the block instance in use.
+ *
+ * It looks up the setting first in the block instance hierarchy.
+ * If none is found, it'll look it up in the block editor settings.
+ *
+ * @param {string} path The path to the setting.
+ * @return {any} Returns the value defined for the setting.
+ * @deprecated
```
Comment (tyxla) [inline packages/block-editor/src/components/use-settings/index.js]: Should we add the version here, like in other places where we use the `@deprecated` tag?

File: packages/block-editor/src/components/use-settings/test/index.js [hunk 22]
```diff
@@ -38,7 +44,19 @@ const mockCurrentBlockContext = (
...
+		const result = hookCb();
+		useEffect( () => {
+			storedResult = result;
+		}, [ result ] );
+	}
+	render( <TestHook /> );
+	return storedResult;
+}
+
+describe( 'useSettings', () => {
```
Comment (tyxla) [inline packages/block-editor/src/components/use-settings/test/index.js:59]: I'd keep a test for `useSetting()` too, even if it's just a single one.

File: packages/block-library/src/cover/controls.native.js [hunk 23]
```diff
@@ -68,14 +68,9 @@ function Controls( {
 		[ minHeight ]
 	);
 
+	const [ availableUnits ] = useSettings( 'spacing.units' );
 	const units = useCustomUnits( {
```
Comment (tyxla) [inline packages/block-library/src/cover/controls.native.js:72]: Not from this PR, but this appears to be so repetitive! I wonder if we should abstract it to a separate hook with these default args.

File: packages/block-library/src/search/edit.js [hunk 24]
```diff
@@ -125,8 +125,10 @@ export default function SearchEdit( {
 	}
 
 	const colorProps = useColorProps( attributes );
-	const fluidTypographySettings = useSetting( 'typography.fluid' );
-	const layout = useSetting( 'layout' );
+	const [ fluidTypographySettings, layout ] = useSettings(
+		'typography.fluid',
+		'layout'
```
Comment (tyxla) [inline packages/block-library/src/search/edit.js:130]: Should we get `layout.wideSize` directly here since that's the only one we need?

File: packages/block-library/src/spacer/edit.native.js [hunk 25]
```diff
@@ -39,10 +39,11 @@ const Spacer = ( {
...
 	};
 	const { height, width } = attributes;
-	const spacingSizes = [
-		{ name: 0, slug: '0', size: 0 },
-		...( useSetting( 'spacing.spacingSizes' ) || [] ),
-	];
+	const spacingSizes = [ { name: 0, slug: '0', size: 0 } ];
+	const [ settingsSizes ] = useSettings( 'spacing.spacingSizes' );
+	if ( settingsSizes ) {
+		spacingSizes.push( ...settingsSizes );
```
Comment (tyxla) [inline packages/block-library/src/spacer/edit.native.js:45]: Would be nice if we could avoid mutation.

File: packages/block-editor/src/components/use-settings/index.js [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+ * @example
+ * ```js
+ * const isEnabled = useSetting( 'typography.dropCap' );
+ * ```
+ */
+export function useSetting( path ) {
+	deprecated( 'wp.blockEditor.useSetting', {
```
Comment (tyxla) [inline packages/block-editor/src/components/use-settings/index.js:264]: At the same time, I'm not convinced `useSetting()` really needs to be deprecated. It uses `useSettings()` under the hood, so what difference does it make if we continue using it when we want to retrieve a single setting?

**Review by tyxla — APPROVED** [review]
This looks great 🚀 Nice optimization!

I've left a few suggestions, but it's minor stuff and nothing really feels blocking.

I'm not particularly convinced that we should deprecate `useSetting()`, I'd rather expect that we discourage sibling `useSetting()` calls and encourage using `useSettings()` in those cases, but this is something we could explore separately.

File: packages/block-editor/src/components/inner-blocks/index.js (hunk 17 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/inner-blocks/index.js:101]: That's not needed because `defaultLayout` is used only with an object spread operator, `...defaultLayout`, which treats `undefined` or `null` as an empty object already.

File: packages/block-editor/src/components/spacing-sizes-control/hooks/use-spacing-sizes.js (hunk 18 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/spacing-sizes-control/hooks/use-spacing-sizes.js:16]: I think that local mutation is fine, it's not worse in any way than creating a new array and concatenating existing arrays into it with the spread operator.

File: packages/block-editor/src/components/use-settings/index.js (hunk 19 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-settings/index.js:170]: Certainly not, it would be counterproductive. If you want to modify a setting with a plugin, you want to register only one filter, not two of them, both `useSetting` and `useSettings`.

File: packages/block-editor/src/components/use-settings/README.md (hunk 20 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-settings/README.md:3]: Probably not, it's sufficiently documented in the jsdoc comments and the top-level README that's generated from them.

File: packages/block-library/src/cover/controls.native.js (hunk 23 above)
Comment (jsnajdr) [reply packages/block-library/src/cover/controls.native.js:72]: Yes! The default units change from case to case, but the rest is the same.

File: packages/block-library/src/search/edit.js (hunk 24 above)
Comment (jsnajdr) [reply packages/block-library/src/search/edit.js:130]: Maybe, but I'll keep the existing `'layout'` retrieval. The codebase consistently uses `'layout'`, never its subfields.

File: packages/block-editor/src/components/use-settings/index.js (hunk 26 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-settings/index.js:264]: Retrieving a single setting is very ergonomic even with `useSettings`. Having also `useSetting` doesn't add any interesting improvement. Therefore, it's a good idea to support only one canonical API, and clearly mark the other one as "inferior".

File: packages/block-editor/README.md (hunk 16 above)
Comment (jsnajdr) [reply packages/block-editor/README.md:948]: Note added :+1:

File: packages/block-editor/src/components/use-settings/index.js (hunk 21 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-settings/index.js]: Version and note added :+1:

File: packages/block-editor/src/components/use-settings/test/index.js (hunk 22 above)
Comment (jsnajdr) [reply packages/block-editor/src/components/use-settings/test/index.js:59]: Added a test for `useSetting` :+1:

**Review by tyxla — APPROVED** [review]
Thanks again @jsnajdr this is a really cool optimization 🚀

**Comment by jsnajdr** [comment]
This PR delivered a significant improvement in performance tests, namely for the "Type" and "Inserter Hovering Items" stats. Immediately visible in the charts on [codevitals.run](https://www.codevitals.run):

Screenshot 2023-10-20 at 17 08 05
Screenshot 2023-10-20 at 17 08 25

**Comment by fabiankaegy** [comment]
This should have been called out in the Misc Dev note for 6.4 as it deprecates a stabile hook

**Comment by Mamaduka** [comment]
I don't remember this being cherry-picked for 6.4. Based on the Gutenberg version this hasn't shipped with latest WP.

**Comment by youknowriad** [comment]
Just added the dev note label to avoid forgetting about it for 6.5

**Comment by jsnajdr** [comment]
I just downloaded WordPress 6.4.1 and inspected the `wp-content/js/dist/block-editor.js` file, and it still has the old version of `useSetting` and no `useSettings`. The patch wasn't backported, it's only in the Gutenberg plugin and will become part of Core only in 6.5.

**Comment by tyxla** [comment]
Yes, 6.4 shipped with Gutenberg up to 16.7, so anything that's in 16.8 or newer and wasn't specifically cherry-picked didn't make it into the release.

**Comment by fabiankaegy** [comment]
Thanks all for the confirmation and sorry for jumping to conclusions too early ❤️

Appreciate you all!

**Comment by fabiankaegy** [comment]
Should we update the version number in the deprecation just for greater clarity?

I created a PR for it here: https://github.com/WordPress/gutenberg/pull/56377

File: packages/block-editor/src/utils/object.js [hunk 27]
```diff
@@ -112,6 +113,8 @@ export function setImmutably( object, path, value ) {
 	return newObject;
 }
 
+const stringToPath = memoize( ( path ) => path.split( '.' ) );
```
Comment (ellatrix) [inline packages/block-editor/src/utils/object.js:116]: This has the drawback of taking up memory that is never released. Maybe harmless here because it's small strings that probably get reused, but still worth nothing. I guess this is a weird case where a max size doesn't make sense. See also https://github.com/WordPress/gutenberg/pull/53406/files#r1286726012

Comment (jsnajdr) [reply packages/block-editor/src/utils/object.js:116]: Removing in #56711. I remember when profiling, the memoization didn't have any impact on performance anyway.

File: packages/block-editor/README.md [hunk 28]
```diff
@@ -962,6 +964,26 @@ _Returns_
...
+### useSettings
+
+Hook that retrieves the given settings for the block instance in use.
+
+It looks up the settings first in the block instance hierarchy. If none are found, it'll look them up in the block editor settings.
+
+_Usage_
+
+```js
+const [ fixed, sticky ] = useSettings( 'position.fixed', 'position.sticky' );
```
Comment (ellatrix) [inline packages/block-editor/README.md:976]: I'm a bit confused here because the PR description says

```js
const [ allowFixed, allowSticky ] = useSettings( [ 'position.fixed', 'position.sticky' ] );
```

Can both be used? It would be nice to be consistent so we don't need to normalise.

Comment (jsnajdr) [reply packages/block-editor/README.md:976]: Only the `useSettings( 'a', 'b' )` form can be used. The PR originally started with an array argument, `useSettings( settings: string[] )`, but then was changed to `useSettings( ...settings: string[] )` during review. That's why the PR description is out of date.

The main reason for that was that when reading a single setting, `useSetting( 'a' )` and `useSettings( 'a' )` are the same, except that for `useSettings` the return value is always an array (with one element).

**Comment by jsnajdr** [comment]
## Dev note

## New `useSettings` hook for reading block instance settings

When trying to improve performance of the block editor for WordPress 6.5, one of the identified issues was related to the `useSetting` hook that is used by block instances to read various settings provided by the environment they are in: parent blocks, the theme, the block editor itself. It turns out that it's inefficent to read multiple settings with separate `useSetting` calls:
```js
const allowFixed = useSetting( 'position.fixed' );
const allowSticky = useSetting( 'position.sticky' );
```
The editor performance is improved when all the settings are read at once, using a new hook called `useSettings`:
```js
const [ allowFixed, allowSticky ] = useSettings( 'position.fixed', 'position.sticky' );
```
That's why WordPress 6.5 introduces this new hook. The `useSetting` (singular) hook is now deprecated (using it will trigger a console warning), because reading even a single setting is very easy with the `useSettings` (plural) hook:
```js
const [ allowFixed ] = useSettings( 'position.fixed' );
```
The only change is that the `useSettings` hook always returns an array, and the array needs to be destructured to read the single value. Supporting two APIs for performing the same task is therefore not justified.

**Comment by ellatrix** [comment]
Makes me wonder if we should just expose the `getBlockSettings` selector too.

**Comment by jsnajdr** [comment]
> we should just expose the `getBlockSettings` selector too.

Yes, of course, let's do that! After all, `useSettings` is just a very thin wrapper around this. And the selector call could be merged into existing `useSelect` calls, reducing the number of store subscriptions.
